main.go

```diff
@@ -1221,28 +1225,151 @@ func populatePackages(currentPackage string, onlyUpdates bool, onlyLatest bool,
+// Before overriding the icons, it will check if the necessary conditions are met at parsePackageListToPackageIconList() function, if not it will skip the package.
+func overrideIcons(c *cli.Context) {
+	currentPackage := os.Getenv(packageEnvVariable)
+	iconOverride := true
+	icons.CheckFilesStructure() // stop execution if file structure is not correct
+
+	// populate all possible packages
+	packageList, err := populatePackages(currentPackage, false, false, false, iconOverride)
+	if err != nil {
+		logrus.Fatal(err)
+	}
+
+	// parse only the packages that have the necessary conditions for icon override
+	packageIconList := parsePackageListToPackageIconList(packageList)
+
+	err = overwriteIndexIconsAndTestChanges(packageIconList)
+	if err != nil {
+		logrus.Errorf("Failed to overwrite index icons: %v", err)
+	}
+
+	err = commitChanges(packageList, iconOverride)
+	if err != nil {
+		logrus.Fatal(err)
+	}
+}
+
+// parsePackageListToPackageIconList will parse the PackageList to PackageIconList
+// and check if the necessary override icon conditions are met
+func parsePackageListToPackageIconList(packageList PackageList) icons.PackageIconList {
+	var packageIconList icons.PackageIconList
+	for _, pkg := range packageList {
+		iconURL := pkg.UpstreamYaml.ChartYaml.Icon
+
+		// check conditions for icon override and avoid panics
+		iconIsDownloaded := icons.CheckForOverrideConditions(pkg.Path, iconURL)
+		if !iconIsDownloaded {
+			logrus.Errorf("Override conditions not met for icon: %s, at path: %s", iconURL, pkg.Path)
+			continue
+		}
+
+		pkgIcon := icons.ParsePackageToOverride(pkg.Name, pkg.Path, iconURL)
+		packageIconList = append(packageIconList, pkgIcon)
+	}
+	return packageIconList
+}
+
+// overwriteIndexIconsAndTestChanges will overwrite the index.yaml icon fields with the new downloaded icons path
+func overwriteIndexIconsAndTestChanges(packageIconList icons.PackageIconList) error {
+	indexFilePath := filepath.Join(getRepoRoot(), indexFile)
+	if _, err := os.Stat(indexFilePath); os.IsNotExist(err) {
+		return err
+	}
+
+	helmIndexYaml, err := repo.LoadIndexFile(indexFilePath)
+	if err != nil {
+		return err
+	}
+	assetsDirectoryPath := filepath.Join(getRepoRoot(), repositoryAssetsDir)
+	newHelmIndexYaml, err := repo.IndexDirectory(assetsDirectoryPath, repositoryAssetsDir)
+	if err != nil {
+		return err
+	}
+	helmIndexYaml.Merge(newHelmIndexYaml)
+	helmIndexYaml.SortEntries()
+
+	icons.ConvertYamlForIconOverride(helmIndexYaml, packageIconList)
+
+	err = helmIndexYaml.WriteFile(indexFilePath, 0644)
+	if err != nil {
+		return err
+	}
+
+	updatedHelmIndexFile, _ := repo.LoadIndexFile(indexFilePath)
+
+	return icons.TestIconsAndIndexYaml(packageIconList, updatedHelmIndexFile)
+}
+
 // func generateChanges(genpatch bool, save bool, commit bool, onlyUpdates bool, print bool) {
-func generateChanges(auto bool, stage bool) {
+func generateChanges(auto bool, stage bool, iconOverride bool) {
```

Comment:
I think that the `iconOverride` argument is also unnecessary here. `generateChanges` is called in three places (I'll leave it to you to find them in your editor):
1. `generateChanges(false, false, false)`
2. `generateChanges(false, true, true)`
3. `generateChanges(true, false, true)`

The only place that `iconOverride` is referenced is line 1365, which contains an if statement evaluating `(auto || stage) && !iconOverride`. If you look at the combinations of arguments that `generateChanges` is used with, you'll notice that this expression never evaluates to `true`.

So why do we have `iconOverride`? Either there is a bug somewhere or we don't need to add the `iconOverride` argument to `generateChanges`.

Reply:
Agreed, reverted